Post: Make PHPStan aware of your api implications (#35)
staabm/staabm.github.io: _posts/2022-12-29-phpstan-analyzable-api-contracts.md

---
tags:
- PHPStan

image: "images/og-images/phpstan-analyzable-api-contracts.jpg"

ogImage:
  title: "Make PHPStan aware of your api implications"
  imageUrl: "https://staabm.github.io/staabm.svg"
  fileName: "phpstan-analyzable-api-contracts"
---

> It's of really great help, when your api contracts are covered by static analysis.

Author: ```suggestion
It's of really great help, when your APIs' contracts are covered by static analysis.
```

## Make PHPStan aware of your APIs' implications️

From time to time I have to implement APIs like this:

```php

class HttpSession {
    /**
     * checks whether the current user has a session
     */ 
    public function sessionExists(): bool
    {
        // implement me
		return (bool) rand(0,1);
    }
    
    /**
     * Returns the current user session if it exists. otherwise returns null. 
     * @return UserSession|null
     */
    public function getSession(): ?UserSession
    {
        // implement me
		return rand(0,1) ? new UserSession() : null;
    }
}

class UserSession {
    // implement me
    function doSomething(): void {}
}

```

> In this case the 2 API methods are kind of interconnected. When a session exists `getSession` will return a `UserSession`.

Author: ```suggestion
In this case the 2 API methods are kind of interconnected. When `sessionExists` returns `true`, `getSession` will return a `UserSession` - in other words it won't return `null`.
```

> From a API consumer point of view I can call `getSession` and check whether the return value is null, or use the dedicated `sessionExists` method.

Author: ```suggestion
From a API consumer point of view I can call `getSession` and check whether the return value is `null`, or use the dedicated `sessionExists` method.
```
